Future::flatten_to_stream

stepancheg opened the pull request.

alexcrichton commented on src/flatten_to_stream.rs:
```diff
@@ -0,0 +1,101 @@
+#[must_use = "streams do nothing unless polled"]
+pub struct FlattenToStream<F>
+    where
+        F : Future,
```
Typically we put this on the same line as `where` above, and omit the space before `:`

stepancheg replied:
Fixed.

alexcrichton commented on src/flatten_to_stream.rs:
```diff
@@ -0,0 +1,101 @@
+}
+
+/// Stream that emits single error
+enum FailedState<T, E> {
```
Does this have to have `T` as a type parameter?

alexcrichton replied:
Also, could this be folded into the state below to avoid nested enums?

stepancheg replied:
> Does this have to have `T` as a type parameter?

To implement `Stream` for `FailedState`, yes.

It is not strictly necessary to implement `Stream` (`T` could be a type parameter of `poll` function), but it makes almost no difference in number of lines of code.

> Also, could this be folded into the state below to avoid nested enums?

Technically, yes, but it would significantly compilcate a loop in `FlattenToStream`:

```
fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
    loop {
        self.state = match self.state {
            ...
            State::Stream(ref mut s) => {
                // This branch relies on the fact that
                // `self.state` remains unchanged after return.
                // So I can't (easily) do `mem::replace(&mut self.state, ..)`
                // with some dummy value.
                return s.poll();
            }
            State::Error(e) => {
                // This branch needs to take ownership of `e` to return it
                // but rust does not allow one match branch to take a reference
                // and other branch to take a value.
                ...
            }
        };
    }
```

alexcrichton replied:
For the impl of `Stream` yeah, it's required, but it seems a little odd here? Also yes you can't have ownership differ on arms of `match`, but that's dealt with in a variety of ways throughout the futures repos.

alexcrichton commented on src/flatten_to_stream.rs:
```diff
@@ -0,0 +1,101 @@
+
+    fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
+        loop {
+            self.state = match &mut self.state {
```
I believe you can avoid the leading `&mut` here and just bind `ref mut` in the bindings below

stepancheg replied:
Done.

alexcrichton commented on src/lib.rs:
```diff
@@ -731,6 +733,38 @@ pub trait Future {
+    /// use futures::stream::Stream;
+    ///
+    /// let future_of_a_stream = finished::<_, u32>(stream::iter(vec![
+    ///         Ok(17), Err(18), Ok(19)].into_iter()));
```
I think the `into_iter` may not be necessary here? Also, could this initialization be split over two lines to avoid the wrapping?

stepancheg replied:
> I think the into_iter may not be necessary here?

It is necessary, because signature of `iter` is:

```
pub fn iter<I, T, E>(i: I) -> IterStream<I>
    where I: Iterator<Item=Result<T, E>>
{ ... }
```

`I` is not `IntoIterator`.

alexcrichton commented on src/flatten_to_stream.rs:
```diff
@@ -0,0 +1,101 @@
+enum FlattenToStreamState<F>
+    where
+        F : Future,
+        <F as Future>::Item: Stream<Error=F::Error>,
```
Could this be `F::Item: Stream<Error=F::Error>`?

stepancheg replied:
`F::Item: Stream<Error=F::Error>` works in beta and nightly, but [complains about cycles](https://ci.appveyor.com/project/alexcrichton/futures-rs/build/1.0.392) in stable.

stepancheg replied:
Updated the patch, because `iter` signature change is merged into master.

alexcrichton commented on src/flatten_to_stream.rs:
```diff
@@ -0,0 +1,101 @@
+    }
+}
+
+enum FlattenToStreamState<F>
```
It should be fine to just call this `State`

stepancheg replied:
Renamed.

stepancheg commented:
> I wonder if we can come up with a perhaps more ergonomic name? Something like `streamed` or similar?

I don't like `flatten_to_stream` name either, but I like `flat` part of it, because it emphasises the fact that it converts a collection of collections into a collection as in `flatten` or `flat_map` on iterators.

Moreover, I think the better purpose of `streamed` identifier is for a combinator that [converts a future into single element stream](https://github.com/stepancheg/grpc-rust/blob/master/src/futures_misc/future_to_stream_once.rs).

stepancheg commented:
Updated the patch, addresses all the issues except:
* name is kept `flatten_to_stream`
* `FailedState` is kept separate enum

(see comments above)

alexcrichton commented:
Perhaps this could be called `into_stream`?

alexcrichton commented:
Hm I guess you could conceptually think of any future as a stream of one element, which arguably is the most generic way to think about futures/streams and what one might expect `into_stream` to do instead, so maybe not.

I still feel that `flatten_to_stream` is a bit long here, but I agree that keeping the name "flat" is probably good as it's what's happening here.

Hm @aturon or @carllerche, any ideas on naming here?

stepancheg commented:
Updated the patch to store a state of `FlattenToStream` in single `State` enum.

Waiting for decision about function name.

aturon commented:
Maybe just `flatten_stream`? I agree that `flatten` is the right word here.

alexcrichton commented:
We could possibly fit in the word "flat" as well. For example `Iterator::flat_map` exists but I opted for `Future::map` + `Future::flatten`. I'm not sure if there's a verb after "flat" though to convey "this is becoming a stream". `flat_stream` is all I can think of but I like that less than `flatten_stream`

aturon commented:
I think this method is niche enough that a longer name is ok, and `flatten_stream` fits the purpose pretty well. (Dropping the `_to_` helps a lot with verbosity imo)

alexcrichton commented:
Agreed, @stepancheg want to rename?

stepancheg commented:
`flatten_stream` is OK with me too. Renamed.

alexcrichton commented:
Thanks!
